sched/semaphore/sem_holder.c

```diff
@@ -1035,17 +1011,52 @@ void nxsem_release_holder(FAR sem_t *sem)
 {
   FAR struct tcb_s *rtcb = this_task();
   FAR struct semholder_s *pholder;
+  FAR struct semholder_s *candidate = NULL;
+  unsigned int total = 0;
 
   /* Find the container for this holder */
 
-  pholder = nxsem_findholder(sem, rtcb);
-  if (pholder != NULL && pholder->counts > 0)
+#if CONFIG_SEM_PREALLOCHOLDERS > 0
+  for (pholder = sem->hhead; pholder; pholder = pholder->flink)
+    {
+#else
+  int i;
+
+  /* We have two hard-allocated holder structures in sem_t */
+
+  for (i = 0; i < 2; i++)
+    {
+      pholder = &sem->holder[i];
+#endif
+
+      if (pholder->counts <= 0)
+        continue;
+
+      if (pholder->htcb == rtcb)
+        {
+          /* Decrement the counts on this holder -- the holder will be freed
+           * later in nxsem_restore_baseprio.
+           */
+
+          pholder->counts--;
+          return;
+        }
+
+      total++;
+      if (candidate == NULL)
+        candidate = pholder;
+    }
+
+  /* The current task is not a holder */
+
+  if (total == 1)
```

Comment:
I'm not sure in this logic. We run this code even if `rtcb` does not match `pholder->htcb`. I'm out of laptop till 8 Jan, so will appreciate if you can describe more a use case that meets this situation.

Reply:
If the caller of sem_post is not a holder (such as task1 call sem_wait and task2 call sem_post), nxsem_findholder(sem, rtcb) must return NULL. 

If the semphore's value >= 2, and more than 1 tasks have called sem_wait, the semphore has multi holders.

If the semphore has only one holder, we can decrement the counts immediately.
Else, it's a bit of a hassle, holders can only be freed until the task exits or the sem destroys.

Reply:
But `sem_post` can't feed more than `SEM_VALUE_MAX`, so `pholder->counts` is limited to same value implicitly, isn't it?

Reply:
I see. You are trying to handle the case when task that posts semaphore never becomes a holder. As an alternative we might limit `pholders->count` by `SEM_VALUE_MAX` before we increment.

Reply:
Yes, the sum of holder's counts must <= SEM_VALUE_MAX, but there may be multi holders. If the caller of sem_post is not a holder, how we choose the holder to decremnt it's counts?

Reply:
> Yes, the sum of holder's counts must <= SEM_VALUE_MAX, but there may be multi holders. If the caller of sem_post is not a holder, how do we choose the holder to decrement it's counts? 

No, the sum of holder's counts is not limited by `SEM_VALUE_MAX`. For example the task1 always call `sem_post` and task2 and task3 try to take semaphore with concurrency to each other. Then sem counts will be incremented and decremented why task2 and task3 holders counts will increment only (currently with rollover).

As it was discussed in the mailing list signaling use case for semaphores is not a valid case, so we do not have any expectations to priority inheritance in this case. So I really think that might think of only limiting holders count by `SEM_VALUE_MAX` and do not decrement at all.

As an alternative to your proposed changes it is better to fing a holder with the highest counts value and use it as a candidate.

Reply:
> If the caller of sem_post is not a holder (such as task1 call sem_wait and task2 call sem_post), nxsem_findholder(sem, rtcb) must return NULL.
> 

in this case, the caller use sem as event, not lock, he should disable priority inheritance first. Why we need take care the wrong usage? it's caller responsibility to fix his bug.